### PR TITLE
Issue #1132 Use /etc/hosts instead of dnsmasq

### PR DESCRIPTION
**Pull request by praveenkumar**
**Fixes:** Issue #1132 

## Solution/Idea

Idea here to remove the dependency on host side dns changes and rely on the `/etc/hosts` file.

## Proposed changes

List main as well as consequential changes you introduced or had to introduce.

There shouldn't be any changes the way we start the crc but now by default you can't open the web browser until this we add to the `/etc/hosts`, ( We need to add console and registry url along with api and outh which is why it is WIP).

## Testing

What is the _bottom-line_ functionality that needs testing? Describe in pseudo-code or in English. Use verifiable statements that tie your changes to existing functionality.

1. `start` succeeds first time after `setup` succeeded

Also this is right now only tested on linux, this will work on the mac since that we are already using but we need more work for windows.

### CODE REVIEW

**Comment by cfergeau** [comment]
This is missing several "big picture" details to fully grasp what this is doing :) Some of the commits (maybe later ones?) will need to detail *why* we are doing this change. Why are we using a separate binary rather than maintaining `/etc/hosts` ourselves? What will the UI like be when the user deploys a new application and wants to access it?

**Comment by jeffsaremi** [comment]
Hi @praveenkumar @gbraad 
This is on Windows and I'm not sure if this change will take care of it or not?
Currently the installation process removes the list of default DNS servers on the "crc" external V-switch. We find this very intrusive. Is there a way through this change or other options to avoid this behavior?

thanks

**Comment by gbraad** [comment]
We are looking into this. It is best to discuss on the main issue associated, the spike about the removal/change. We are gathering details about what this should look like

**Comment by gbraad** [comment]
Also see #926

**Comment by cfergeau** [comment]
> This is missing several "big picture" details to fully grasp what this is doing :) Some of the commits (maybe later ones?) will need to detail _why_ we are doing this change. Why are we using a separate binary rather than maintaining `/etc/hosts` ourselves? What will the UI like be when the user deploys a new application and wants to access it?

There is still no good overview of what the overall goal of this series is, how things are expected to work after it, why we are making this change, ...

File: pkg/crc/services/dns/dns_linux.go [hunk 1]
```diff
@@ -1,12 +1,36 @@
...
+	for _, hostname := range hostnames {
+		if _, _, err := crcos.RunWithDefaultLocale(goodhostPath, "rm", hostname); err != nil {
+			return err
+		}
+		if _, _, err := crcos.RunWithDefaultLocale(goodhostPath, "add", instanceIP, hostname); err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (cfergeau) [inline pkg/crc/services/dns/dns_linux.go:39]: Can the changes in `dns_linux.go` be put in a separate commit?

File: pkg/crc/services/dns/dns_linux.go [hunk 2]
```diff
@@ -1,12 +1,36 @@
...
+		serviceConfig.BundleMetadata.GetAppHostname("oauth-openshift")); err != nil {
+		result.Success = false
+		return *result, err
+	}
 	result.Success = true
 	return *result, nil
 }
+
+// updateHostsConfFile updates the host's /etc/hosts file with Instance IP.
+func updateHostsConfFile(instanceIP string, hostnames ...string) error {
```
Comment (cfergeau) [inline pkg/crc/services/dns/dns_linux.go:27]: Maybe just `updateHostsFile`?

File: pkg/crc/services/dns/dns_linux.go [hunk 3]
```diff
@@ -12,7 +12,10 @@ func runPostStartForOS(serviceConfig services.ServicePostStartConfig, result *se
 	// We might need to set the firewall here to forward
 	// Update /etc/hosts file for host
 	if err := updateHostsConfFile(serviceConfig.IP, serviceConfig.BundleMetadata.GetAPIHostname(),
-		serviceConfig.BundleMetadata.GetAppHostname("oauth-openshift")); err != nil {
+		serviceConfig.BundleMetadata.GetAppHostname("oauth-openshift"),
+		serviceConfig.BundleMetadata.GetAppHostname("console-openshift-console"),
+		serviceConfig.BundleMetadata.GetAppHostname("default-route-openshift-image-registry"),
+		serviceConfig.BundleMetadata.GetAppHostname("foo")); err != nil {
```
Comment (cfergeau) [inline pkg/crc/services/dns/dns_linux.go:18]: Why is this linux specific? We'll want exactly the same thing on all platforms.
I don't think we should add `foo` here, but rather we should use the test for `foo` to indicate whether dnsmasq-based DNS is working or not. This will be useful when debugging networking issues.

File: pkg/crc/services/dns/dns_darwin.go [hunk 4]
```diff
@@ -31,7 +31,9 @@ type resolverFileValues struct {
 func runPostStartForOS(serviceConfig services.ServicePostStartConfig, result *services.ServicePostStartResult) (services.ServicePostStartResult, error) {
 	// Update /etc/hosts file for host
 	if err := updateHostsConfFile(serviceConfig.IP, serviceConfig.BundleMetadata.GetAPIHostname(),
-		serviceConfig.BundleMetadata.GetAppHostname("oauth-openshift")); err != nil {
+		serviceConfig.BundleMetadata.GetAppHostname("oauth-openshift"),
+		serviceConfig.BundleMetadata.GetAppHostname("console-openshift-console"),
+		serviceConfig.BundleMetadata.GetAppHostname("default-route-openshift-image-registry")); err != nil {
```
Comment (cfergeau) [inline pkg/crc/services/dns/dns_darwin.go:36]: There seems to be a mix of
1) change the existing code to make use of goodhosts-cli to update /etc/hosts
2) add console-openshift-console/default-route-openshift-image-registry to /etc/hosts

I would split this commit along these lines, and merge 1) with the next commit in this series ("Remove `/etc/hosts` file modification on MacOS"). Or replace this with a commit using the same code on linux and macos.

File: pkg/crc/preflight/preflight_checks_common.go [hunk 5]
```diff
@@ -104,3 +111,24 @@ func fixPodmanBinaryCached() error {
...
+}
+
+func fixGoodHostBinaryCached() error {
+	goodhostPath := filepath.Join(constants.CrcBinDir, constants.GoodhostBinaryName)
+	goodhost := goodhost.GoodhostCached{}
+	if err := goodhost.EnsureIsCached(); err != nil {
+		return fmt.Errorf("Unable to download goodhost remote binary %v", err)
+	}
+	logging.Debug("goodhost remote binary cached")
+	return setSuid(goodhostPath)
```
[the remark this answers is not in the document]
Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_common.go:133]: shouldn't the binary be embedded?

File: pkg/crc/goodhost/goodhost_cache.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+		return false
+	}
+	return true
+}
+
+func matchGoodhostBinaryName(filename string) bool {
+	return filepath.Base(filename) == constants.GoodhostBinaryName
+}
+
+func (goodhost *GoodhostCached) getgoodhost(destDir string) (string, error) {
```
[the remark this answers is not in the document]
Comment (gbraad) [reply pkg/crc/goodhost/goodhost_cache.go:42]: It should be Goodhosts to reflect the name of the library

File: pkg/crc/cache/cache.go [hunk 7]
```diff
@@ -48,7 +48,7 @@ func CacheBinary(cache Cache, matchBinaryName func(string) bool) error {
 
 	for _, extractedFilePath := range extractedFiles {
 		finalBinaryPath := filepath.Join(outputPath, filepath.Base(extractedFilePath))
-		err = crcos.CopyFileContents(extractedFilePath, finalBinaryPath, 0500)
+		err = crcos.CopyFileContents(extractedFilePath, finalBinaryPath, 0555)
```
Comment (cfergeau) [inline pkg/crc/cache/cache.go:51]: Maybe this should be done by `setSUID` instead?

File: pkg/crc/cache/cache.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	"path/filepath"
+
+	"github.com/code-ready/crc/pkg/crc/constants"
+	"github.com/code-ready/crc/pkg/extract"
+	crcos "github.com/code-ready/crc/pkg/os"
+	"github.com/pkg/errors"
+)
+
+type Cache interface {
+	GetBinary(destDir string) (string, error)
```
[the remark this answers is not in the document]
Comment (cfergeau) [reply pkg/crc/cache/cache.go]: True, `IsCached` is still needed, but can't this be all generic code? I'm not sure an interface is needed, just a type with the appropriate methods should do the trick?

File: pkg/extract/extract.go [hunk 9]
```diff
@@ -139,10 +139,6 @@ func unzip(archive, target string, fileFilter func(string) bool) ([]string, erro
 			return nil, fmt.Errorf("%s: illegal file path", path)
 		}
 
-		if fileFilter != nil && !fileFilter(path) {
-			logging.Debugf("untar: Skipping %s", path)
-			continue
-		}
```
[the remark this answers is not in the document]
Comment (cfergeau) [reply pkg/extract/extract.go:142]: The current code is
```
// the target location where the dir/file should be created
		path := filepath.Join(targetDir, header.Name)
		if fileFilter != nil && !fileFilter(path) {
			logging.Debugf("untar: Skipping %s", path)
			continue
		}

		// check the file type
		switch header.Typeflag {
		// if it's a file create it
		case tar.TypeReg, tar.TypeGNUSparse:
			// tar.Next() will externally only iterate files, so we might have to create intermediate directories here
			if err = os.MkdirAll(filepath.Dir(path), 0750); err != nil {
				return err
			}
			file, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, header.FileInfo().Mode())
			if err != nil {
				return err
			}
			defer file.Close()

			// copy over contents
			if _, err := io.Copy(file, tarReader); err != nil {
				return err
			}
		}
```
There is a call to `mkdirAll(path)` before trying to create the file from the archive, is this not working?

File: pkg/crc/cache/cache.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (cfergeau) [reply pkg/crc/cache/cache.go]: This could be squashed in the appropriate commits for example https://github.com/cfergeau/crc/commit/1adb910b1214995e97eff5a532aa8cfea7046ead

[the remark this answers is not in the document]
Comment (cfergeau) [reply pkg/crc/cache/cache.go]: On second though, `OcCache`, `PodmanCache`, ... are not needed so this can become https://github.com/cfergeau/crc/commit/3b1d32143a1ba749ab8130a1a7503eb737444819

[the remark this answers is not in the document]
Comment (praveenkumar) [reply pkg/crc/cache/cache.go]: @cfergeau we also need to include the path because right now we are dumping everything in `~/.crc/bin` so this will again need a refactor when https://github.com/code-ready/crc/pull/1217 worked on.

File: pkg/extract/extract.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (praveenkumar) [reply pkg/extract/extract.go:142]: @cfergeau no this not going to work, because if you check the current code the `path` here is got from `path := filepath.Join(targetDir, header.Name)` and `header.Name` only contain the file name not the complete path so in our current scenario we check first the `fileFilter(path)` and skip that directory creation if the binary is inside a sub-directory. 

Assume binary name is `oc` which we want to extract and it is available in the tarball like `A/B/oc` so as per current flow when tar.Next() hit the `B` then fileFilter check if the `B` is `oc` if not then continue without creating the directory `B` and then when tar.Next() counter to `oc` it identify the binary is `oc` and now it hit the ` os.MkdirAll(filepath.Dir(path), 0750)` where the `path` will only have `<target_dir>/oc` instead `<target_dir>/B/oc` and it will fail.

**Comment by praveenkumar** [comment]
retest this please.

File: pkg/crc/cache/cache.go [hunk 10]
```diff
@@ -23,6 +24,14 @@ func New(binarName string, archiveURL string, destDir string) *Cache {
...
 }
 
+func NewOcCache(destDir string) *Cache {
+	return New(constants.OcBinaryName, constants.GetOcUrl(), destDir)
+}
+
+func NewPodmanCache(destDir string) *Cache {
+	return New(constants.PodmanBinaryName, constants.GetPodmanUrl(), destDir)
+}
+
```
Comment (cfergeau) [inline pkg/crc/cache/cache.go:34]: I think Gerard objected to having this in the generic `Cache` package, this could be moved to one of the preflight files for now.

File: pkg/crc/preflight/preflight_checks_darwin.go [hunk 11]
```diff
@@ -84,7 +84,7 @@ func setSuid(path string) error {
 	}
 
 	/* Can't do this before the chown as the chown will reset the suid bit */
-	stdOut, stdErr, err = crcos.RunWithPrivilege(fmt.Sprintf("set suid for %s", path), "chmod", "u+s", path)
+	stdOut, stdErr, err = crcos.RunWithPrivilege(fmt.Sprintf("set suid for %s", path), "chmod", "u+s,g+x", path)
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_darwin.go]: Hmm this assumes that the file is u+x, this assumes the user running crc is part of the `wheel` group. Not sure we want to address this right now, but `setSuid` could probably still fail in some corner cases.

File: pkg/crc/cache/cache.go [hunk 12]
```diff
@@ -32,6 +32,10 @@ func NewPodmanCache(destDir string) *Cache {
 	return New(constants.PodmanBinaryName, constants.GetPodmanUrl(), destDir)
 }
 
+func NewGoodhostCache(destDir string) *Cache {
+	return New(constants.GoodhostBinaryName, constants.GetGoodHostUrl(), destDir)
```
Comment (cfergeau) [inline pkg/crc/cache/cache.go:36]: `Goodhost` or `GoodHost`? And this really should be `Goodhosts` or `GoodHosts`  (applies throughout the PR)

File: pkg/crc/preflight/preflight_checks_common.go [hunk 13]
```diff
@@ -26,6 +26,12 @@ var genericPreflightChecks = [...]PreflightCheck{
 		fixDescription:   "Caching podman remote binary",
 		fix:              fixPodmanBinaryCached,
 	},
+	{
+		checkDescription: "Checking if goodhosts binary is cached",
+		check:            checkGoodhostBinaryCached,
+		fixDescription:   "Caching goodhosts binary",
+		fix:              fixGoodHostBinaryCached,
+	},
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_common.go:34]: Do we want  to make it mandatory by default?

File: pkg/crc/preflight/preflight_checks_common.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cfergeau) [reply pkg/crc/preflight/preflight_checks_common.go:121]: I think at the moment it's not being embedded.

File: pkg/crc/services/dns/dns_darwin.go (hunk 4 above)
Comment (cfergeau) [reply pkg/crc/services/dns/dns_darwin.go:38]: This also duplicates the linux code.

File: pkg/crc/services/dns/dns_darwin.go [hunk 14]
```diff
@@ -144,35 +146,14 @@ func waitForNetwork() error {
...
-	}
-	return nil
-}
-
-func replaceHostname(hosts *goodhosts.Hosts, ipAddress string, hostname string) error {
-	logging.Debugf("Updating %s to %s in /etc/hosts file", hostname, ipAddress)
-	if hosts.HasHostname(hostname) {
-		if err := hosts.RemoveByHostname(hostname); err != nil {
+		if _, _, err := crcos.RunWithDefaultLocale(goodhostPath, "add", instanceIP, hostname); err != nil {
 			return err
```
Comment (cfergeau) [inline pkg/crc/services/dns/dns_darwin.go:155]: Would probably be nicer (more readable) to have a helper to add/remove an entry to /etc/hosts.

File: pkg/crc/cache/cache.go (hunk 12 above)
Comment (praveenkumar) [reply pkg/crc/cache/cache.go:36]: @cfergeau I am going to use `Goodhost` throughout the PR.

File: pkg/crc/preflight/preflight_checks_common.go (hunk 13 above)
Comment (praveenkumar) [reply pkg/crc/preflight/preflight_checks_common.go:34]: @cfergeau I am incline to make it mandatory and since this is on our crc-repo side so we have full control over it.

File: pkg/crc/preflight/preflight_checks_common.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (praveenkumar) [reply pkg/crc/preflight/preflight_checks_common.go:121]: @cfergeau it is part of embed.go in the same PR so it is going to be embedded.

File: pkg/crc/cache/cache.go (hunk 10 above)
Comment (gbraad) [reply pkg/crc/cache/cache.go:34]: I mind to have the values IN the cache package as that makes it non-generic.
if they exist and get set outside the cache they would make this fiel really generic and reusable without having to modify this file... and just from constansts or the preflight that uses it

**Comment by praveenkumar** [comment]
> I'm still missing a clear description of what is the exact goal with this change. Is it just a workaround for when we fail to configure DNS so that the cluster at least starts?

@cfergeau Yes, that is the initial plan for this PR.

>  Do we plan for this to replace DNS based configuration?

This PR doesn't replace the existing DNS based configuration but just act as an addon so if the DNS configuration fail then at least cluster start and user can able to login using cli or webconsole.

>What are the plans for adding/removing additional entries for user apps?

At this moment adding/removing user app or any other route is not scope of this PR, in future we can have a `crc route add/remove` command for this (but as I said this is not scope for this PR and shouldn't block for this usecase.)

**Comment by cfergeau** [comment]
> > Do we plan for this to replace DNS based configuration?
> 
> This PR doesn't replace the existing DNS based configuration but just act as an addon so if the DNS configuration fail then at least cluster start and user can able to login using cli or webconsole.
> 

My feeling is that we are basically moving the failure from `crc start` to a later point when the user is trying to create an application. I guess we should at least add a warning that we are running in a degraded state, and that application routes are not going to automatically work?

File: pkg/crc/cache/cache.go (hunk 12 above)
Comment (cfergeau) [reply pkg/crc/cache/cache.go]: `git grep  goodhost[^s]` still returns some results

File: pkg/crc/cache/cache.go (hunk 10 above)
Comment (cfergeau) [reply pkg/crc/cache/cache.go:34]: This does not seem to be addressed in the latest iteration? Is there nothing to address?

File: pkg/crc/preflight/preflight_checks_common.go (hunk 13 above)
Comment (cfergeau) [reply pkg/crc/preflight/preflight_checks_common.go:34]: I realized it can't be skipped at all, so we'll always have the 2 DNS configuration methods running in parallel. Enabling it by default or not is one thing, but I really think it should be possible to disable it entirely, at least for debugging/testing purposes. Then you will have people who don't want (can't) run a 5.5MB suid binary.

**Comment by cfergeau** [comment]
> > > Do we plan for this to replace DNS based configuration?
> > 
> > 
> > This PR doesn't replace the existing DNS based configuration but just act as an addon so if the DNS configuration fail then at least cluster start and user can able to login using cli or webconsole.
> 
> My feeling is that we are basically moving the failure from `crc start` to a later point when the user is trying to create an application. I guess we should at least add a warning that we are running in a degraded state, and that application routes are not going to automatically work?

Overall the patch series looks good wrt achieving its goal. However I'm really dubious about not being able to disable this at all, especially as it will make bugs in the 'main' implementation harder to notice, harder to debug, ...
